### PR TITLE
#2654 add lookup record creation thunks

### LINKED ISSUE #2654: Move patient API network logic to thunks
## Is your feature request related to a problem? Please describe.

Lookup requests are currently called directly from the `SearchForm` component. To be consistent with the mobile architecture, these should be encapsulated in thunks.
 
## Describe the solution you'd like

Add redux logic for patient API network requests.

## Implementation

Most consistent solution is to keep the lookup modal state in the store, and do the heavy lifting in the reducer. At the moment, this might be a bit more work that it's worth. Suggested solution is to keep the parameters in the component state, and inject them into the reducer via the action payload.

## Describe alternatives you've considered

N/A.

## Additional context

See epic #2446.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #2654.

## Change summary

Update existing patient and prescriber actions and migrate patient API.

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] New patients are created successfully.
- [ ] Existing patients are updated successfully.
- [ ] Patients added via the API are created successfully.
- [ ] New prescribers are created successfully.
- [ ] Existing prescribers are updated successfully.
- [ ] Prescribers added via the API are created successfully.

### Related areas to think about

N/A.

### CODE REVIEW

File: src/actions/PatientActions.js [hunk 1]
```diff
@@ -27,27 +27,94 @@ const editPatient = patient => ({
...
+    billAddress3,
+    billAddress4,
+    billPostalZipCode,
+    country,
+    female,
+    supplyingStoreId,
+    isActive,
+  };
+
+  UIDatabase.write(() => createRecord(UIDatabase, 'Patient', patientRecord));
```
Comment (joshxg) [inline src/actions/PatientActions.js:113]: I find code like this very interesting - you want to be explicit, but that's so the code is clear and easy to read, but then if it's too long and explicit, it becomes harder to read!

I haven't checked every field but I think something a little more JS could make it a little easier on the eyes

```
const record = { ...currentPatient, ...{... patientDetails, suppliyingStore: UID..(thisstore), name: ... },  
```

File: src/actions/PrescriberActions.js [hunk 2]
```diff
@@ -32,24 +32,70 @@ const editPrescriber = prescriber => ({
...
+    addressTwo,
+    phoneNumber,
+    mobileNumber,
+    emailAddress,
+    female,
+    storeId,
+    isActive,
+  };
+
+  UIDatabase.write(() => createRecord(UIDatabase, 'Prescriber', prescriberRecord));
```
Comment (joshxg) [inline src/actions/PrescriberActions.js:94]: same as above - quite easy to merge objects with JS, so I think taking advantage of that while using comments to describe the objects shape makes for the cleanest code here!

File: src/database/utilities/createRecord.js [hunk 3]
```diff
@@ -17,15 +17,20 @@ import { SETTINGS_KEYS } from '../../settings';
...
+ * @param   {string}        id        Id of the address (can be undefined).
  * @param   {string}        line1     Line 1 of the address (can be undefined).
  * @param   {string}        line2     Line 2 of the address (can be undefined).
  * @param   {string}        line3     Line 3 of the address (can be undefined).
  * @param   {string}        line4     Line 4 of the address (can be undefined).
  * @param   {string}        zipCode   Zip code of the address (can be undefined).
  * @return  {Realm.object}            The Address object described by the params.
  */
-export const getOrCreateAddress = (database, line1, line2, line3, line4, zipCode) => {
+export const getOrCreateAddress = (database, id, line1, line2, line3, line4, zipCode) => {
```
Comment (joshxg) [inline src/database/utilities/createRecord.js:28]: I believe this is going to break the calls in `incomingSyncUtils.js`? Might want to make it the last param? possibly having an object param, but quite a few changes over just making it the last param

File: src/database/utilities/createRecord.js [hunk 4]
```diff
@@ -132,36 +137,44 @@ const createAddress = (database, { line1, line2, line3, line4, zipCode } = {}) =
...
 
   const id = prescriberId ?? generateUUID();
-  const thisStoreId = database.getSetting(SETTINGS_KEYS.THIS_STORE_ID);
-  const supplyingStoreId = prescriberStoreId ?? thisStoreId;
-  const fromThisStore = supplyingStoreId === thisStoreId;
+  const firstName = prescriberFirstName ?? '';
+  const lastName = prescriberLastName ?? '';
+  const registrationCode = prescriberRegistrationCode ?? '';
+
+  const address = getOrCreateAddress(database, addressOne, addressTwo);
```
Comment (joshxg) [inline src/database/utilities/createRecord.js:164]: wrong params?

File: src/database/utilities/createRecord.js (hunk 3 above)
Comment (wlthomson) [reply src/database/utilities/createRecord.js:28]: Ah, forgot to make the changes to the other callers, plan was to just add a `null` param, but that's pretty ugly. Tbh, I think longer term this method can be made a bit nicer, but I think for now making an object param is nicer, it's not called in that many places!

File: src/actions/PatientActions.js (hunk 1 above)
Comment (wlthomson) [reply src/actions/PatientActions.js:113]: Yeah, should have been more descriptive in the PR. For now I'm keeping it explicit to make it clear exactly what params are coming from where and going where. The probably with the nice spread syntax is that you can't easily see what parameters are actually being used (I had to cross-reference other files and it was easier in the end just to log everything).  

To answer the question re. explicitness being harder to read, I think key point is that explicitness isn't to make the code nicer or easier to read (it's definitely harder to read and uglier!), the point is to show the data being passed around. IMO, in this case using spread syntax makes it nicer to look at, but doesn't make it easier to read, as the information is all hidden, so it has to be found by reading other methods and logging etc. 

Saying all that, the plan is still to make it nice and tidy once all the reviewing and stuff is done (i.e. at the point where functionality is working well and there's less need for that data to be explicit).

File: src/actions/PrescriberActions.js (hunk 2 above)
Comment (wlthomson) [reply src/actions/PrescriberActions.js:94]: As above, using spread syntax with comments is in the pipeline for the tidyup stage, but I think a bit premature at this point.

File: src/database/utilities/createRecord.js (hunk 3 above)
Comment (wlthomson) [reply src/database/utilities/createRecord.js]: cfafe1f96cbfb8f91bc811b521b2b9c81542954b.

File: src/database/utilities/createRecord.js (hunk 4 above)
Comment (wlthomson) [reply src/database/utilities/createRecord.js]: cfafe1f96cbfb8f91bc811b521b2b9c81542954b.

**Pull request merged**